Source/WTF/wtf/EnumeratedArray.h

```diff
@@ -0,0 +1,264 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <array>
+
+namespace WTF {
+
+// This is an std::array where the indices of the array are values of an enum (rather than a size_t).
+// This assumes the values of the enum start at 0 and monotonically increase by 1
+// (so the conversion function between size_t and the enum is just a simple static_cast).
+// N is the maximum value of the enum, which determines the size of the array.
+template <typename Key, typename T, Key N>
```

Comment:
Might be slightly confusing that `N` here is not the count, but the last value. Maybe name this `LastValue`?

It's also different from `std::array` which does take a count, but that can't be helped if we don't want to add a final `Count` enum value to our enums, and we assume that writing things like `EnumeratedArray<MyEnum, ..., static_cast<MyEnum>(static_cast<size_t>(MyEnum::LastValue) + 1)>` is too awkward (which it is).

Reply:
Yeah, you're exactly right.